[EDU-2448] refactor: cache keys into edge caching (#463)
aziontech/docs: src/content/docs/en/pages/main-menu/reference/build/edge-application/edge-caching.mdx

---

## Cache keys

> A cache key is an index entry for an object in Azion's edge nodes. When a first request is made to an edge application, the edge requests the objects within the request to the origin server. This generates a cache for each object at the edge. When the next request for the same resource is made to the edge application, if the resource requested matches one already found through the cache key, the request won't be forwarded to the origin.

Author: ```suggestion
A cache key is an index entry for an object in Azion's edge nodes. When a request is made to an edge application for the first time, the edge requests the objects within the request to the origin server. This generates a cache for each object at the edge. When the next request for the same resource is made to the edge application, if the resource requested matches one already found through the cache key, the request won't be forwarded to the origin.
```

You can obtain the cache key of an application's resource through a request for the content URI and add the Azion debug header `Pragma: azion-debug-cache`:

```bash
curl -I https://yourdomain.com/resource-path/image.jpeg -H "Pragma: azion-debug-cache"
```

A successful response will return an `X-Cache-Key` header with the cache key value, which based on the example above, would be:

```
X-Cache-Key: httpsyourdomain.com/resource-path/image.jpeg@@
```

:::note
Find out more about how to obtain other cache data in the [How to verify application cache indicators using Modheader for Google Chrome](/en/documentation/products/guides/check-page-cache-time/) guide.
:::

### Cache key format

The default cache key format adopted by Azion concatenates the following URI syntax elements:

> - Scheme of the request

Author: ```suggestion
- Scheme of the request (for complex requests)
```

- Host
- Path to the resource

> - Query string

Author: ```suggestion
- Query string
- A variation separator and appended variations
```

---

## Cache keys

A cache key is an index entry for an object in Azion's edge nodes. When a request is made to an edge application for the first time, the edge forwards the request to the origin server and generates a cache for each object received at the edge. When the next request for the same resource is made to the edge application, if the resource requested matches one already found through the cache key, the request won't be forwarded to the origin.

> You can obtain the cache key of an application's resource through a `HEAD` request for the content URI and add the Azion debug header `Pragma: azion-debug-cache`:

Reviewer: The request doesn't have to be of HEAD method. Any method (GET, OPTIONS, etc) does return pragma headers. The only exception is when the return is an error (4xx, 5xx)

Author: Left general statement!

> You can obtain the cache key of an application's resource through a `HEAD` request for the content URI and add the Azion debug header `Pragma: azion-debug-cache`:

Author: ```suggestion
You can obtain the cache key of an application's resource through a request for the content URI and add the Azion debug header `Pragma: azion-debug-cache`:
```

```bash

> curl --location --head 'https://yourdomain.com/resource-path/image.jpeg' \

Reviewer: --head is not mandatory.

Author: Simplified the command and explanation 😉

> curl --location --head 'https://yourdomain.com/resource-path/image.jpeg' \
> --header 'Pragma: azion-debug-cache'

Author: ```suggestion
curl -I https://yourdomain.com/resource-path/image.jpeg -H "Pragma: azion-debug-cache"
```

```

> The response will return an `X-Cache-Key` header with the cache key value, which based on the example above, would be:

Author: ```suggestion
A successful response will return an `X-Cache-Key` header with the cache key value, which based on the example above, would be:
```

```
X-Cache-Key: httpsyourdomain.com/resource-path/image.jpeg@@
```

:::note
Find out more about how to obtain other cache data in the [How to verify application cache indicators using Modheader for Google Chrome](/en/documentation/products/guides/check-page-cache-time/) guide.
:::

### Cache key format

The default cache key format adopted by Azion concatenates the following URI syntax elements:

> - Scheme of the request (for complex requests)

Author: ```suggestion
- Scheme of the request
```

- Host
- Path to the resource

> - Query string

Author: ```suggestion
- Query strings
```

> - A variation separator and appended variations

Author: ```suggestion
- A variation separator and, when implemented, appended variations
```

An object with the URI `https://static.yourdomain.com/page/site.js` should generate the cache key `httpsstatic.yourdomain.com/page/site.js`.

> Some cache keys may also change depending on the following variations generated at the edge.

Author: ```suggestion
Some cache keys may also change depending on variations generated at the edge. If variations are enabled, the cache key may contain the `@@` separator at the end, resulting in `httpsstatic.seudominio.com/pagina/site.js@@`.
```

#### Complex requests

> If the resource is obtained using a complex request (any method different than `GET` or `HEAD`), the method is added to the cache key. For example, for an `OPTIONS` request, the cache key generated for an object would be `optionshttpsstatic.yourdomain.com/page`.

Author: ```suggestion
If the resource is obtained using a complex request (any method different than `GET` or `HEAD`), the method is added to the start of the cache key.

For example, for an `OPTIONS` request, the cache key generated for an object would be `optionshttpsstatic.yourdomain.com/page`.
```

#### Image Processor

> When [Image Processor](/en/documentation/products/edge-application/image-processor/) is activated and an image is resized and converted into `webp` format, the cache key generated will be `httpsstatic.yourdomain.com/static/images/image_1.jpg?ims=880x@@webp`.

Author: ```suggestion
When [Image Processor](/en/documentation/products/edge-application/image-processor/) is activated, each `ims` query element is separated by the separator.

For example, for an image that is resized and converted into `webp` format, the cache key generated will be `httpsstatic.yourdomain.com/static/images/image_1.jpg?ims=880x@@webp`.
```

```

:::note
Find out more about how to obtain other cache data in the [How to verify application cache indicators using Modheader for Google Chrome](/en/documentation/products/guides/check-page-cache-time/) guide.
:::

### Cache key format

The default cache key format adopted by Azion concatenates the following URI syntax elements:

- Scheme of the request
- Host
- Path to the resource
- Query strings
- A variation separator and, when implemented, appended variations

An object with the URI `https://static.yourdomain.com/page/site.js` should generate the cache key `httpsstatic.yourdomain.com/page/site.js`.

Some cache keys may also change depending on variations generated at the edge. If variations are enabled, the cache key may contain the `@@` separator at the end, resulting in `httpsstatic.seudominio.com/pagina/site.js@@`.

#### Complex requests

If the resource is obtained using a complex request (any method different than `GET` or `HEAD`), the method is added to the start of the cache key.

For example, for an `OPTIONS` request, the cache key generated for an object would be `optionshttpsstatic.yourdomain.com/page`.

#### Image Processor

> When [Image Processor](/en/documentation/products/edge-application/image-processor/) is activated, each `ims` query element is separated by the separator.

Author: ```suggestion
When [Image Processor](/en/documentation/products/edge-application/image-processor/) is activated, in addition to the `ims` query string, variations with converted format contain the file format after the separator.
```

For example, for an image that is resized and converted into `webp` format, the cache key generated will be `httpsstatic.yourdomain.com/static/images/image_1.jpg?ims=880x@@webp`.

#### Device Groups

> If there's an object variation per [device group](/en/documentation/products/edge-application/device-groups/), the cache key will be appended with the separator `@@` and the device group name; for example: `httpwww.yourdomain.com/@@Mobile`.

Author: ```suggestion
If there's an object variation per [device group](/en/documentation/products/edge-application/device-groups/), the cache key will be appended with the separator `@@` and the device group name

For example, for an application that implements variation for the `Mobile` group, the generated cache key will be `httpwww.yourdomain.com/@@Mobile`.
```

#### Advanced Cache Key

When the [Advanced Cache Key](/en/documentation/products/edge-application/application-acceleration/#advanced-cache-key) feature is configured, the formation of cache keys is also affected. 

> **Cache by Cookie** generates a cache key variation with the determined cookie names and values followed by `;`. For example, content variation based on the `user` cookie could generate the following cache keys:

Author: ```suggestion
**Cache by Cookie** generates a cache key variation with the determined cookie names and values followed by `;`. 

For example, content variation based on the `user` cookie could generate the following cache keys:
```

- `httpwww.yourdomain.com/@@;`
- `httpwww.yourdomain.com/@@user=user;`

> **Cache by Query String** generates cache keys with the query separator and the determined arguments based on the order of the query strings submitted in the request. For example, content variation based on `name` and `city` could generate the following cache keys:

Author: ```suggestion
**Cache by Query String** generates cache keys with the query separator and the determined arguments based on the order of the query strings submitted in the request. 

For example, content variation based on `name` and `city` could generate the following cache keys:
```

- `httpstatic.yourdomain.com/page?name=name`
- `httpstatic.yourdomain.com/page?city=city&name=name`
- `httpstatic.yourdomain.com/page?name=name&city=city`

For multiple query strings, if **Query string sort** is enabled, the query strings will be grouped under the same cache key, ordered alphabetically.

#### File slicing

> For sliced objects, the cache key appends the `@@bytes=` separator for each slice of content. Cache keys using optimization for delivering large files have the following format:

Author: ```suggestion
For sliced objects, the cache key appends the `@@bytes=` separator for each slice of content.

For example, for a file that is `2097151` bytes in size, if file slicing is enabled, the cache keys generated will be:
```

- `httpstatic.yourdomain.com/media/file.mp4@@bytes=0-1048575`
- `httpstatic.yourdomain.com/media/file.mp4@@bytes=1048576-2097151`

#### Cached HTTP methods

> For cached `POST` or `OPTIONS` requests, the cache key appends the `@@` separator followed by the MD5 hash of the request body. For example:

Author: ```suggestion
For cached `POST` or `OPTIONS` requests, the cache key appends the `@@` separator followed by the MD5 hash of the request body. 

For example:
```
